### ISSUE TITLE
pretraining parameters loading problem about CE loss training

### ISSUE THREAD

**Issue by jack1yang** [ruotianluo/self-critical.pytorch#154]
Thanks for your code first. 
Step1: follow your instructions, i first train model using cross-entropy about 30 epochs,  using the following commond
`python train.py --id fc --caption_model fc --input_json data/cocotalk.json --input_fc_dir data/cocotalk_fc --input_att_dir data/cocotalk_att --input_label_h5 data/cocotalk_label.h5 --batch_size 10 --learning_rate 5e-4 --learning_rate_decay_start 0 --scheduled_sampling_start 0 --checkpoint_path log_fc --save_checkpoint_every 6000 --val_images_use 5000 --max_epochs 30`

Step2: then using self-critical to finetune the model to get a higher cider, using the following command:
(i'm so sorry to write  the wrong commands at the first time:`python scripts/prepro_ngrams.py --input_json .../dataset_coco.json --dict_json data/cocotalk.json --output_pkl data/coco-train --split train`)
In fact , I want to say:
`python train.py --id fc_rl --caption_model fc --input_json data/cocotalk.json --input_fc_dir data/cocotalk_fc --input_att_dir data/cocotalk_att --input_label_h5 data/cocotalk_label.h5 --batch_size 10 --learning_rate 5e-5 --start_from log_fc_rl --checkpoint_path log_fc_rl --save_checkpoint_every 6000 --language_eval 1 --val_images_use 5000 --self_critical_after 30 --cached_tokens coco-train-idxs`

however,  i not found your code loading the model's parameters trained from step 1 using cross-entropy loss. Because I found that  Self-critical Sequence Training for Image Captioning says that we should first train the model using cross-entropy with some epochs and then finetune it with self-critical. However,  i only found that 
`# Load the optimizer
    if vars(opt).get('start_from', None) is not None and os.path.isfile(os.path.join(opt.start_from,"optimizer.pth")):
        optimizer.load_state_dict(torch.load(os.path.join(opt.start_from, 'optimizer.pth')))`,
 not found like this command,
`torch.load(mode.pth)`
so, when training model with self-critical, we just need to initialize the model's parameters with random values instead of initializing from pre-training model using cross-entropy?

**Comment by Jordan-5i** [comment]
You may run the script  scripts/copy_model.sh， this command is copy the model of training cross-entropy in new folder with the tagger of  ‘_rl’ , like fc_rl or adaatt_rl etc. Then run `python train.py --id fc_rl --caption_model fc --input_json data/cocotalk.json --input_fc_dir data/cocotalk_fc --input_att_dir data/cocotalk_att --input_label_h5 data/cocotalk_label.h5 --batch_size 10 --learning_rate 5e-5 --start_from log_fc_rl --checkpoint_path log_fc_rl --save_checkpoint_every 6000 --language_eval 1 --val_images_use 5000 --self_critical_after 30 --cached_tokens coco-train-idxs`  in which the --id is the the model training in reinforce learning and --cached_tokens is the result of running  the Step2 in your comments.

**Comment by jack1yang** [comment]
> You may run the script scripts/copy_model.sh， this command is copy the model of training cross-entropy in new folder with the tagger of ‘_rl’ , like fc_rl or adaatt_rl etc. Then run `python train.py --id fc_rl --caption_model fc --input_json data/cocotalk.json --input_fc_dir data/cocotalk_fc --input_att_dir data/cocotalk_att --input_label_h5 data/cocotalk_label.h5 --batch_size 10 --learning_rate 5e-5 --start_from log_fc_rl --checkpoint_path log_fc_rl --save_checkpoint_every 6000 --language_eval 1 --val_images_use 5000 --self_critical_after 30 --cached_tokens coco-train-idxs` in which the --id is the the model training in reinforce learning and --cached_tokens is the result of running the Step2 in your comments.

Thanks for you reply first!!! I known you provided commands. However,  I’m still not entirely sure how train.py loads pre-trained models. PyTorch’s documentation recommends torch.load() or torch.load_state_dict() — I see torch.load_state_dict() used for the optimizer, but neither used for the main model or dp_model variables.
I also see infos = cPickle.load(f) & histories = cPickle.load(f) which seem to resemble torch.load(), but the infos and histories variables don’t seem to be used to influence the model or dp_model variables. How are the weights loaded into the model or dp_model variables?

**Comment by jack1yang** [comment]
> You may run the script scripts/copy_model.sh， this command is copy the model of training cross-entropy in new folder with the tagger of ‘_rl’ , like fc_rl or adaatt_rl etc. Then run `python train.py --id fc_rl --caption_model fc --input_json data/cocotalk.json --input_fc_dir data/cocotalk_fc --input_att_dir data/cocotalk_att --input_label_h5 data/cocotalk_label.h5 --batch_size 10 --learning_rate 5e-5 --start_from log_fc_rl --checkpoint_path log_fc_rl --save_checkpoint_every 6000 --language_eval 1 --val_images_use 5000 --self_critical_after 30 --cached_tokens coco-train-idxs` in which the --id is the the model training in reinforce learning and --cached_tokens is the result of running the Step2 in your comments.

Thanks again! I found the command that load the pre-trained model's parameters. It is in  models/__init__.py() 
`model.load_state_dict(torch.load(os.path.join(opt.start_from,` `'model.pth')))`

**Comment by ruotianluo** [comment]
FYI, it's now moved to train.py
